src/v/cloud_topics/reader/placeholder_extent.cc

```diff
@@ -0,0 +1,315 @@
+      std::move(header_bytes));
+    // NOTE: the serialized raft_data batch doesn't have the offset set
+    // so we need to populate it from the placeholder batch. We also need
+    // to make sure that crc is correct.
+    header.base_offset = base_offset;
+    header.crc = model::crc_record_batch(header, records_bytes);
+    crc::crc32c crc;
+    model::crc_record_batch_header(crc, header);
+    header.header_crc = crc.value();
+    model::record_batch batch(
+      header,
+      std::move(records_bytes),
+      model::record_batch::tag_ctor_ng{}); // TODO: fix compression
+    return batch;
+}
+
+template<class Clock>
+ss::future<result<bool>> basic_placeholder_extent<Clock>::materialize(
+  cloud_storage_clients::bucket_name bucket,
+  cloud_io::remote_api<Clock>* api,
+  cloud_io::basic_cache_service_api<Clock>* cache,
+  basic_retry_chain_node<Clock>* rtc) {
+    bool hydrated = false;
+    // This iobuf contains the record batch replaced by the placeholder. It
+    // might potentially contain data that belongs to other placeholder
+    // batches and in order to get the extent of the record batch
+    // placeholder we need to use byte offset and size.
+    iobuf L0_object_content;
+
+    // 2. download object from S3
+    auto cache_file_name = std::filesystem::path(
+      ssx::sformat("{}", placeholder.id()));
+    // TODO: replace this with proper object name
+    // currently this value is used as both cloud storage name
+    // and cache name. This shouldn't necessary be the case in the
+    // future.
+
+    std::optional<cloud_io::cache_element_status> status = std::nullopt;
+    basic_retry_chain_node<Clock> is_cached_rtc(retry_strategy::backoff, rtc);
+    retry_permit rp = is_cached_rtc.retry();
+    while (rp.is_allowed && !status.has_value()) {
+        auto is_cached_result
+          = result_from_ready_future<errc::cache_read_error>(
+            co_await ss::coroutine::as_future(
+              cache->is_cached(cache_file_name)));
+
+        if (!is_cached_result.has_value()) {
+            co_return is_cached_result.error();
+        }
+
+        switch (is_cached_result.value()) {
+        case cloud_io::cache_element_status::available:
+        case cloud_io::cache_element_status::not_available:
+            status = is_cached_result.value();
+            break;
+        case cloud_io::cache_element_status::in_progress:
+            // Another fiber is trying to put value into the cache.
+            // Wait until the operation is completed but stay within the
+            // time budget.
+            if (rp.abort_source != nullptr) {
+                co_await ss::sleep_abortable(rp.delay, *rp.abort_source);
+            } else {
+                co_await ss::sleep<Clock>(rp.delay);
+            }
+            rp = is_cached_rtc.retry();
+            continue;
+        }
+    }
+
+    if (!rp.is_allowed) {
+        co_return errc::timeout;
+    }
+
+    if (
+      !status.has_value()
```

Comment:
@Lazin

Reply:
will clean up in a followup, this check is not needed because if `status.has_value() == false` we will return timeout error from previous if statement (the only way status could be a nullopt is when the loop timed out while cache element is in-progress, in this case rp.is_allowed will be equal to false).